### PR TITLE
Binary and Decimal Conversion Components.

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
Added a couple components that convert a number into an array of binary digits and viceversa.

## Why It's Good For The Game
May help players mak more complex and advanced components without filtering those lacking knowledge in the binary field.

## Changelog

:cl:
expansion: Added binary and decimal conversion components.
/:cl:

### CODE REVIEW

**Comment by Ghommie** [comment]
Please remove the mechanic label and add the content expansion one

**Comment by optimumtact** [comment]
why shouldn't they be filtered? that's what skill progression is

**Comment by Watermelon914** [comment]
> why shouldn't they be filtered? that's what skill progression is

it's more for convenience as well, you don't want to pull out your calculator whenever you need to do any sort of bit math.

**Comment by Ghommie** [comment]
> 
> 
> why shouldn't they be filtered? that's what skill progression is

Watermelon told me circuits components should be novice-friendly on discord.

**Comment by CRITAWAKETS** [comment]
also making an entire bit array for a compact remote is a pain

**Comment by optimumtact** [comment]
convenience in a game is boring, it kills features

**Comment by Watermelon914** [comment]
for something as simple as this, I don't think it should be very hard to approach as someone who's never done computer science before. 

There are other components that filter out those who may not be experienced enough to use them, but this is one of those things that if done badly becomes unusable, unappealing and hard to learn to anyone who hasn't learnt how bits in binary work and it becomes too inconvenient to those who do know that using a method such as checking for numbers in a list literal becomes more convenient.

File: code/modules/wiremod/components/abstract/binary_decimal.dm [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	return
+
+/obj/item/circuit_component/binary_decimal/populate_ports()
+	set_array_size(default_array_size)
+
+/obj/item/circuit_component/binary_decimal/Destroy()
+	number = null
+	bit_array = null
+	return ..()
+
```
Comment (Watermelon914) [inline code/modules/wiremod/components/abstract/binary_decimal.dm:72]: you don't need to do this

File: code/modules/wiremod/components/abstract/binary_decimal.dm [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		var/datum/port/output = bit_array[index]
+		bit_array -= output
+		remove_bit_port(output)
+		array_size--
+
+	while(array_size < new_size)
+		array_size++
+		var/index = length(bit_array)
+		if(!trigger_input)
+			index += 1
```
Comment (Watermelon914) [inline code/modules/wiremod/components/abstract/binary_decimal.dm:56]: remove these lines

File: code/modules/wiremod/components/math/binary_conversion.dm [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	display_name = "Binary Conversion"
+	desc = "Splits a decimal number into an array of binary digits, or bits, represented as 1 or 0 and often used in boolean or binary operations like AND, OR and XOR. \
+		Attack in hand to increase array size, right click to decrease array size."
+
+/obj/item/circuit_component/binary_decimal/binary_conversion/populate_ports()
+	. = ..()
+	number = add_input_port("Number", PORT_TYPE_NUMBER)
+
+/obj/item/circuit_component/binary_decimal/binary_conversion/add_bit_port(index)
+	add_output_port("Bit [index]", PORT_TYPE_NUMBER)
```
Comment (Watermelon914) [inline code/modules/wiremod/components/math/binary_conversion.dm:16]: this doesn't really add to any list?

**Review by Watermelon914 — CHANGES_REQUESTED** [review]
I think this PR is flawed because it has to copy a lot of code to get things done. I'm going to DNR this PR until I can look at turning the behaviour used in list literals into some sort of component-like thing so that you don't have to copy and paste code every time you want to be able to increase/decrease the amount of ports in a circuit component.

I'll probably get that done today day or tomorrow plus extra for however long it takes to get merged.

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

**Pull request merged**